### ISSUE TITLE
Does simd_mask need copy_to, copy_from and ctor load?

### ISSUE THREAD

**Issue by danieltowner** [mattkretz/std-simd-feedback#32]
simd_mask provides copy_to and copy_from which read and write memory regions containing bool values. I think this is a fairly unusual operation because the native simd_mask implementation is likely to be a full set of element bits (AVX-like), or a compact bit set (e.g., ARM, AVX-512). The copy_to/copy_from functions do provide a way of generically storing simd_masks to and from memory, but I'm not sure I like forcing them to be a particular format which doesn't match either of the likely common implementations.

Given that simd_mask is proposed to have the operator+ or a conversion operator, could a programmer use that instead to explicitly mimic the effect:

    simd_mask<uint8_t> = some_mask;
    simd<uint8_t>(some_mask).copy_to(uint8_t_ptr);
    // or (+some_mask).copy_to(uint8_t_ptr);

Similarly, reading back from memory can be accomplished using operator! to convert from simd elements to mask bits.

    !(simd<uint8_t>(ptr_to_memory))

I think we should remove those memory operations and require the user to explicitly manage the storage when they need to.

**Comment by mattkretz** [comment]
Yes, this is only a consistency aspect. I believe it does have a value that the interfaces of `simd` and `simd_mask` are as similar as possible.
That said, yes `bool` seems like the worst choice. So if we keep `simd_mask::copy_from`/`to`, do you have a suggestion for a better type? `std::bitset` with deduced width?

**Comment by mattkretz** [comment]
@danieltowner I'm working on P1928R3 for next week. If you want this question discussed in LEWG, I could write something on this topic. Let me know.

**Comment by danieltowner** [comment]
I don't think there is a benefit to being consistent with simd unless it can be done efficiently and with no surprises. I think it would be preferable to remove those functions entirely so that the programmer recognises that it is up to them to decide how to store a mask into memory, whether via a simd<> of their choosing, or a bitset.

If you felt strongly that there is still value to having copy_to/copy_from then I think I would lean towards the bits being contiguous in memory (i.e., like bitset). Care would be needed with regard to how much data would be written - is it byte granularity (i.e., round up to the next byte but no more), or larger (like the natural or most efficient word size)?

**Comment by mattkretz** [comment]
Background on `simd_mask`: https://wg21.link/N4185. The original motivation:

> Mask types can implement load and store functions, reading from / writing to arrays of `EntryType` (which is `bool`). These functions can be useful to write code that is independent of the SIMD register width and to interface with non-SIMD code (or I/O in general). The `Flags` argument is analogous to the one for the `Vector<T>` load/store functions. The default uses unaligned loads and stores and can be set to aligned loads and store via the second argument.

(Note how I originally proposed `std::element_aligned` to be the default... :sigh:)

More questions:
* If/when we make `simd_mask` a range, what type does the iterator dereference return? `bool`, right? So `std::ranges::copy(mask, ptr)` acts equivalent to the current `simd_mask::copy_to`. If we make it a mutable range, same for `copy_from`. Only that the ranges interface will certainly be less efficient.
* If `simd_mask` is convertible to `bitset` and users start dumping those into memory/file/network for I/O and on the other side the `simd_mask` has a different width... I expect many failures, because `bitset` typically pads to 32 or even 64 bits. Shouldn't we help those users?
* The alternative: `static_cast<fixed_size_simd<int8_t, mask.size>>(mask).copy_to(ptr)` and `static_cast<mask_type>(fixed_size_simd<int8_t, mask_type::size>(ptr))`. Do you believe this can be as efficient?
* Or do you believe nobody will do `simd_mask` I/O? I remember at least one user who used mask loads/stores and it didn't seem stupid...

**Comment by danieltowner** [comment]
Deferring the range question because I'm not sure whether it makes sense to provide iterators on simd at all.

 A bool is only a byte because you can't address anything smaller; so while simd_mask might be logically bool/ that doesn't mean we need use bytes for storage. std::vector<bool> compacts bools into bits, as does std::bitset, so I don't think it istoo much of a leap to say that simd_mask can store bools in compact form too.

I do want to help users, but making the only option for copying a mask to memory and back use a storage method which doesn't match the most likely implementation methods doesn't seem helpful. It pushes programmers to use methods which are potentially inefficient.

I think the alternatives could be efficient. Using my library ``(+m).copy_to(ptr)`` and ``simd<uint8_t>(ptr) != 0`` generated good code, and a peephole in the compiler would make some cases better too.

The only places I've encountered masks being stored in memory is when our algorithm is already using some sort of compacted representation, such as plain unsigned values. That is partially why I would like _to_ullong() and constructor from integral, because I know that is the storage format used for a particular wireless application. Those particular functions have the limitation that they store no more than 64 elements, but that could be side-stepped by using _BitInt(n).

To borrow another idea from bitset (I know, not necessarily the best role model :-) we could provide ``to_string`` and a ``bitset(std::string)`` constructor which allows it to be portably copied to and from memory.

**Comment by mattkretz** [comment]
https://wg21.link/P0161R0 "Bitset Iterators, Masks, and Container Operations" is relevant. But the paper has not been pursued further by the author. :frowning_face:

As much as everyone hates `vector<bool>`, do you believe it could make sense to have load & store functions using `std::vector<bool>::iterator` as parameter? This would make it possible to store bits contiguously and independent of `simd_mask::size`.

Sorry for clinging to the load and store functions here, but I still believe consistency is important. It makes learning the SIMD library (or any library) easier if you learn one thing and can extrapolate from there.

OK, now that I proposed iterators for simd_mask load & store, maybe simd should have that as well:
```c++
class simd {
  template <std::contiguous_iterator It>
  requires std::indirectly_writable<It, const value_type&>
  void copy_to(It first, Flags);

  void copy_from(std::contiguous_iterator auto first, Flags);
```

**Comment by danieltowner** [comment]
Yes to the contiguous_iterator idea.

vector<bool> doesn't guarantee contiguous storage otherwise that would have been a solution.
